apps/taxes/src/modules/taxjar/taxjar-calculate-taxes-adapter.ts

```diff
@@ -0,0 +1,141 @@
+import { distributeDiscount } from "../taxes/distribute-discount";
+import { numbers } from "../taxes/numbers";
+import { taxProviderUtils } from "../taxes/tax-provider-utils";
+import { CalculateTaxesResponse } from "../taxes/tax-provider-webhook";
+import { taxJarAddressFactory } from "./address-factory";
+import { FetchTaxForOrderArgs, TaxJarClient } from "./taxjar-client";
+import { TaxJarConfig } from "./taxjar-config";
+import { WebhookAdapter } from "../taxes/tax-webhook-adapter";
+
+type Payload = {
+  taxBase: TaxBaseFragment;
+  channel: ChannelConfig;
+};
+
+type Target = FetchTaxForOrderArgs;
+type Response = CalculateTaxesResponse;
+
+export class TaxJarCalculateTaxesPayloadTransformer {
+  private mapLines(taxBase: Payload["taxBase"]): Target["params"]["line_items"] {
+    const { lines, discounts } = taxBase;
+    const discountSum = discounts?.reduce(
+      (total, current) => total + Number(current.amount.amount),
+      0
+    );
+    const linePrices = lines.map((line) => Number(line.totalPrice.amount));
+    const distributedDiscounts = distributeDiscount(discountSum, linePrices);
+
+    return lines.map((line, index) => {
+      const discountAmount = distributedDiscounts[index];
+
+      return {
+        id: line.sourceLine.id,
+        // todo: replace
+        chargeTaxes: true,
+        // todo: get from tax code matcher
+        taxCode: "",
+        quantity: line.quantity,
+        // todo: clarify if I need to include discount in total amount
+        totalAmount: Number(line.totalPrice.amount),
+        unitAmount: Number(line.unitPrice.amount),
+        discount: discountAmount,
+      };
+    });
+  }
+
+  transform({ taxBase, channel }: Payload): Target {
+    const fromAddress = taxJarAddressFactory.fromChannelAddress(channel.address);
+
+    if (!taxBase.address) {
+      throw new Error("Customer address is required to calculate taxes in TaxJar.");
+    }
+
+    const toAddress = taxJarAddressFactory.fromSaleorAddress(taxBase.address);
+
+    const taxParams = {
+      params: {
+        ...fromAddress,
+        ...toAddress,
+        shipping: taxBase.shippingPrice.amount,
+        line_items: this.mapLines(taxBase),
+      },
+    };
+
+    return taxParams;
+  }
+}
+
+export class TaxJarCalculateTaxesResponseTransformer {
+  private mapLines(res: TaxForOrderRes["tax"]): Response["lines"] {
+    const lines = res.breakdown?.line_items ?? [];
+
+    return lines.map((line) => ({
+      total_gross_amount: taxProviderUtils.resolveOptionalOrThrow(
+        line.taxable_amount + line.tax_collectable,
+        new Error("Taxable amount and tax collectable are required to calculate gross amount")
+      ),
+      total_net_amount: taxProviderUtils.resolveOptionalOrThrow(
+        line.taxable_amount,
+        new Error("Taxable amount is required to calculate net amount")
+      ),
+      tax_rate: line.combined_tax_rate ?? 0,
+    }));
+  }
+
+  private mapShipping(
```

Comment:
you can extract dedicated methods/classes like `TaxjarShippingMapper` and have easier unit tests for these methods.

And this class can call the other class (composition/delegation)

Reply:
Following the current naming convention, the name is: `TaxJarCalculateTaxesResponseShippingTransformer` (the same goes for the file name: `taxjar_calculate_taxes/tax_jar_calculate_taxes_response_shipping_transformer.ts`), where:
- `TaxJar` - the provider name
- `CalculateTaxes - the webhook name
- `Response` - Payload/Response, suggests which way it transforms the data
- `Shipping` - suggests it's about shipping
- `Transformer` - suggests what it does: transforms. Like a lil' adapter

That's a long name. What do you think? Should I at least skip the `taxjar_calculate_taxes` in the filename, since it's in `taxjar_calculate_taxes` folder?